### PR TITLE
docs/node: Add Upgrading Key Managers

### PR DESCRIPTION
**Pull request by amela**
Preview: https://deploy-preview-375--trusting-archimedes-14c863.netlify.app/node/run-your-node/keymanager-node/key-manager-upgrade

### CODE REVIEW

File: docs/node/run-your-node/keymanager-node/key-manager-upgrade.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
+# Upgrading Key Managers
+
+This guide will describe how to upgrade a Key Manager node.
+
+## About the Upgrade
+
+Every Key Manager node contains all the keys used by confidential ParaTimes 
+inside its TEE-encrypted state. The key material is sealed and can only be 
+decrypted by the exactly same TEE enclave running on the exactly same CPU. This
```
Comment (kostko) [inline docs/node/run-your-node/keymanager-node/key-manager-upgrade.md]:
```suggestion
decrypted by exactly the same TEE enclave running on exactly the same CPU. This 
```

File: docs/node/run-your-node/keymanager-node/key-manager-upgrade.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+## About the Upgrade
+
+Every Key Manager node contains all the keys used by confidential ParaTimes 
+inside its TEE-encrypted state. The key material is sealed and can only be 
+decrypted by the exactly same TEE enclave running on the exactly same CPU. This 
+means that newer Key Manager paratimes can not read the key material and that 
+key material can not be restored on another machine. 
+
+**During a Key Manager node upgrade it is therefore essential that the key
```
Comment (kostko) [inline docs/node/run-your-node/keymanager-node/key-manager-upgrade.md]: Use a warning / danger box for this.

File: docs/node/run-your-node/keymanager-node/key-manager-upgrade.md [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+A Key Manager node's upgrade procedure differs from other Oasis nodes upgrades 
+because the upgraded node cannot unseal/decrypt the old Key Manager's state.
+
+To upgrade a Key Manager node, we need to delete the local state and let the 
+Key Manager's state replicate itself from other nodes. Only one Key Manager 
+runtime in the configuration file can be present at once. 
+
+**Always follow the safe upgrade procedure:**
+1. Keep the first half of nodes running on the old version
```
Comment (kostko) [inline docs/node/run-your-node/keymanager-node/key-manager-upgrade.md]: Which half? The guide is about upgrading a single node, so this seems a bit confusing. Should it start out as assuming there are two nodes or 2N nodes to upgrade?

File: docs/node/run-your-node/keymanager-node/key-manager-upgrade.md [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+2. Upgrade the second half
+3. Wait for the paratime upgrade epoch
+4. Verify that secrets have been replicated to the second half of nodes
```
Comment (kostko) [inline docs/node/run-your-node/keymanager-node/key-manager-upgrade.md]: How to verify this? I believe it is mentioned below, so just say something like "(as shown below)" or similar.

File: docs/node/run-your-node/keymanager-node/key-manager-upgrade.md [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+6. Upgrade the first half of the nodes
+
+### Upgrade Nodes
+
+To upgrade a Key Manager, follow the next steps:
+1. Wipe its local state `worker-local-storage.badger.db`
+```
+rm -rf runtimes/4000000000000000000000000000000000000000000000004a1a53dff2ae482d/worker-local-storage.badger.db/
+```
+2. Upgrade the Key Manager runtime
```
Comment (kostko) [inline docs/node/run-your-node/keymanager-node/key-manager-upgrade.md]: Maybe explain what this means (e.g. get the new ORC, update the configuration to replace the ORC, restart the node).

File: docs/node/run-your-node/keymanager-node/key-manager-upgrade.md [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+2. Upgrade the second half
+3. Wait for the paratime upgrade epoch
+4. Verify that secrets have been replicated to the second half of nodes
+5. **Verify again**
+6. Upgrade the first half of the nodes
+
+### Upgrade Nodes
+
+To upgrade a Key Manager, follow the next steps:
+1. Wipe its local state `worker-local-storage.badger.db`
```
Comment (kostko) [inline docs/node/run-your-node/keymanager-node/key-manager-upgrade.md]: Stop the node first.

File: docs/node/run-your-node/keymanager-node/key-manager-upgrade.md [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+    "access_list": [
+...
+    ],
+    "private_peers": [
+...
+    ]
+  }
+```
+
+#### Errors
```
Comment (kostko) [inline docs/node/run-your-node/keymanager-node/key-manager-upgrade.md]: Move this to a top-level "Troubleshooting" section, similar to what we have for the "ParaTime Node" page.

File: docs/node/run-your-node/keymanager-node/key-manager-upgrade.md [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+```
+rm -rf runtimes/4000000000000000000000000000000000000000000000004a1a53dff2ae482d/worker-local-storage.badger.db/
+```
+2. Upgrade the Key Manager runtime
+3. Wait for the key material to get replicated from active nodes before 
+continuing
+
+### After the Upgrade
+
+#### If everything goes well
```
Comment (kostko) [inline docs/node/run-your-node/keymanager-node/key-manager-upgrade.md]: Maybe call it something like "Verifying Successful Replication".

File: docs/node/run-your-node/keymanager-node/key-manager-upgrade.md (hunk 3 above)
Comment (amela) [reply docs/node/run-your-node/keymanager-node/key-manager-upgrade.md]: Changed it a little bit. Hope now is better. I just wanted to say that upgrading all Key Manager nodes at the same time is not a good idea.

File: docs/node/run-your-node/keymanager-node/key-manager-upgrade.md [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+A Key Manager node's upgrade procedure differs from other Oasis nodes upgrades 
+because the upgraded node cannot unseal/decrypt the old Key Manager's state.
+
+To upgrade a Key Manager node, we need to delete the local state and let the 
+Key Manager's state replicate itself from other nodes. Only one Key Manager 
+runtime in the configuration file can be present at once. 
+
+**In case you are running multiple Key Manager nodes always follow the safe 
+upgrade procedure:**
+1. Keep the approximately one half of nodes running on the old version
```
Comment (kostko) [inline docs/node/run-your-node/keymanager-node/key-manager-upgrade.md]:
```suggestion
1. Keep approximately one half of nodes running the old version.
```

File: docs/node/run-your-node/keymanager-node/key-manager-upgrade.md [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+2. Upgrade the other half
+3. Wait for the paratime upgrade epoch
```
Comment (kostko) [inline docs/node/run-your-node/keymanager-node/key-manager-upgrade.md]:
```suggestion
3. Wait for the ParaTime upgrade epoch.
```

File: docs/node/run-your-node/keymanager-node/key-manager-upgrade.md [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+because the upgraded node cannot unseal/decrypt the old Key Manager's state.
+
+To upgrade a Key Manager node, we need to delete the local state and let the 
+Key Manager's state replicate itself from other nodes. Only one Key Manager 
+runtime in the configuration file can be present at once. 
+
+**In case you are running multiple Key Manager nodes always follow the safe 
+upgrade procedure:**
+1. Keep the approximately one half of nodes running on the old version
+2. Upgrade the other half
```
Comment (kostko) [inline docs/node/run-your-node/keymanager-node/key-manager-upgrade.md]:
```suggestion
2. Upgrade the other half.
```

File: docs/node/run-your-node/keymanager-node/key-manager-upgrade.md [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+3. Wait for the paratime upgrade epoch
+4. Verify that secrets have been replicated [as shown below]
```
Comment (kostko) [inline docs/node/run-your-node/keymanager-node/key-manager-upgrade.md]:
```suggestion
4. Verify that secrets have been replicated [as shown below].
```

File: docs/node/run-your-node/keymanager-node/key-manager-upgrade.md [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+5. **Verify again**
```
Comment (kostko) [inline docs/node/run-your-node/keymanager-node/key-manager-upgrade.md]:
```suggestion
5. **Verify again.**
```

File: docs/node/run-your-node/keymanager-node/key-manager-upgrade.md [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+6. Upgrade the rest of the nodes
```
Comment (kostko) [inline docs/node/run-your-node/keymanager-node/key-manager-upgrade.md]:
```suggestion
6. Upgrade the rest of the nodes.
```

File: docs/node/run-your-node/keymanager-node/key-manager-upgrade.md [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+
+[as shown below]: #verifying-successful-replication
+
+### Upgrade Nodes
+
+To upgrade a Key Manager, follow the next steps:
```
Comment (kostko) [inline docs/node/run-your-node/keymanager-node/key-manager-upgrade.md]:
```suggestion
To upgrade a key manager node, follow the next steps:
```

File: docs/node/run-your-node/keymanager-node/key-manager-upgrade.md [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+1. Stop the node
```
Comment (kostko) [inline docs/node/run-your-node/keymanager-node/key-manager-upgrade.md]:
```suggestion
1. Stop the node.
```

File: docs/node/run-your-node/keymanager-node/key-manager-upgrade.md [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+2. Wipe its local state `worker-local-storage.badger.db`
```
Comment (kostko) [inline docs/node/run-your-node/keymanager-node/key-manager-upgrade.md]:
```suggestion
2. Wipe its local state `worker-local-storage.badger.db`, e.g.:
```

File: docs/node/run-your-node/keymanager-node/key-manager-upgrade.md [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+```
+rm -rf runtimes/4000000000000000000000000000000000000000000000004a1a53dff2ae482d/worker-local-storage.badger.db/
+```
+3. Upgrade the Key Manager runtime: 
+    - get the new ORC file ([mainnet], [testnet])
```
Comment (kostko) [inline docs/node/run-your-node/keymanager-node/key-manager-upgrade.md]:
```suggestion
    - get the new ORC file ([mainnet], [testnet]); 
```

File: docs/node/run-your-node/keymanager-node/key-manager-upgrade.md [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+
+To upgrade a Key Manager, follow the next steps:
+1. Stop the node
+2. Wipe its local state `worker-local-storage.badger.db`
+```
+rm -rf runtimes/4000000000000000000000000000000000000000000000004a1a53dff2ae482d/worker-local-storage.badger.db/
+```
+3. Upgrade the Key Manager runtime: 
+    - get the new ORC file ([mainnet], [testnet]) 
+    - update the configuration to replace the ORC file
```
Comment (kostko) [inline docs/node/run-your-node/keymanager-node/key-manager-upgrade.md]:
```suggestion
    - update the configuration to replace the ORC file; and
```

File: docs/node/run-your-node/keymanager-node/key-manager-upgrade.md [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+    - restart the node
```
Comment (kostko) [inline docs/node/run-your-node/keymanager-node/key-manager-upgrade.md]:
```suggestion
    - restart the node.
```

File: docs/node/run-your-node/keymanager-node/key-manager-upgrade.md [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+4. Wait for the key material to get replicated from active nodes before 
+continuing
```
Comment (kostko) [inline docs/node/run-your-node/keymanager-node/key-manager-upgrade.md]:
```suggestion
continuing.
```

File: docs/node/run-your-node/keymanager-node/key-manager-upgrade.md [hunk 22]
```diff
@@ -0,0 +1,118 @@
...
+5. Verify that secrets have been replicated [as shown below]
```
Comment (kostko) [inline docs/node/run-your-node/keymanager-node/key-manager-upgrade.md]:
```suggestion
5. Verify that secrets have been replicated [as shown below].
```

**Review by kostko — COMMENTED** [review]
Nice, looks mostly good, just added some minor stylistic nits.

Also maybe just use lowercase "key manager" everywhere instead of "Key Manager".

**Review by kostko — APPROVED** [review]

**Pull request merged**